### ISSUE TITLE
An inferred actual type argument should be a denotable type

### ISSUE THREAD

**Issue by eernstg** [dart-lang/language#254]
This issue proposes a _principle_ which we could use as a guideline when decisions must be made about how to perform type inference. I believe that we are mostly following this guideline already, but see below for some 'interesting' cases.

**Principle:** If type inference transforms an expression _e<sub>1</sub>_ by adding some actual type arguments _T<sub>1</sub> .. T<sub>k</sub>_, yielding the expression _e<sub>2</sub>_, then it must be true for each _j_ in 1 .. _k_ that there exists syntax which denotes _T<sub>j</sub>_.

The point is that it would be useful to have a clear limit on the expressive power of type inference of the form "you could have written it yourself", which helps keeping programs comprehensible, and helps us avoid situations where we have to say to developers "you must convince type inference to give you that type argument, because there is no way you could write it explicitly".

We have some immediate issues:

- Generic function instantiation will pass type arguments to a generic function _F_, thus yielding a non-generic function whose type is a generic instantiation of the type of _F_. Problem: We do not have the syntactic support to pass those type arguments at all (`f<int>` is not an expression). Solution: We might be able to add support for this syntax, and we have already considered that for a while; so this issue may not be so serious.

- But privacy may be harder. Consider the following example, which shows that we already support giving a variable an inferred type that we cannot denote:

```dart
// Library lib.dart.
class A { foo() { print("foo"); } }
class _B extends A { bar() { print("bar"); } }
_B f() => _B();

// Library main.
import 'lib.dart';

main() {
  var x = f();
  x.foo(); 
  x.bar(); 
}
```

- The same treatment is given in a case where the type is reified, namely for function literal parameters:

```dart
// Library lib.dart.
class A { foo() { print("foo"); } }
class _B extends A { bar() { print("bar"); } }

void f(void Function(_B) g) {
  print(g.runtimeType); // Shows parameter type `_B`.
  g(_B());
}

// Library main.
import 'lib.dart';

main() {
  f((b) { b.foo(); b.bar(); });
}
```

- Considering an example like https://github.com/dart-lang/sdk/issues/36096, we probably want to specify that inference can pass a type argument which is not statically safe (example: passing `num` as the type argument to `printT` in `t2.printT()`, even though the bound is only known to be some subtype of `num`). So this is probably a non-problem, because we will pass `num` or make it an error, and those solutions will satisfy the principle that I'm proposing here.

**Comment by munificent** [comment]
Even without private types, you can get into situations where inference may produce a type you can't write in the current scope:

```dart
// a.dart
class A {}

// b.dart
import 'a.dart';
A makeAnA() => A();

// my_app.dart
import 'b.dart';

main() {
  var a = makeAnA();
}
```

Here, there is no annotation you can write instead of `var` unless you also add an import. I think code like this feels very natural to Dart users (and users of other languages). Private types are a slightly harder one because then there isn't a library you can import, but I think it still roughly "feels the same" for most users — there is a type that *could* be written, you just can't write it *here*.

So I think the principle is too restrictive. I would rephrase it that every type created by inference should be denotable *somewhere* in the program. So a private or un-imported type is OK because you could denote it in that other library. But if, say, inference produced an intersection type and we don't add `&` type syntax, then the principle is violated.

**Comment by eernstg** [comment]
Right, thanks!
